network-libp2p/src/connection_pool/behaviour.rs

```diff
@@ -161,7 +198,16 @@ impl<T: Ord> ConnectionState<T> {
     fn mark_banned(&mut self, id: T) {
         self.failed.remove(&id);
         self.down.remove(&id);
-        self.banned.insert(id);
+        self.banned.insert(id.clone());
+
+        let unban_deadline = Instant::now() + self.ban_time;
+        // If no peer is scheduled to be unbanned, put is as the first-to-be unbanned peer.
+        // Otherwise queue the peer such that it will be scheduled later when the peer is the first in line.
+        if self.unban_timeout.is_none() {
+            self.unban_timeout = Some((id, Box::pin(sleep_until(unban_deadline))));
```

Comment:
You'll need to wake yourself here, so that the `sleep_until` future gets polled at some point.

```suggestion
            self.unban_timeout = Some((id, Box::pin(sleep_until(unban_deadline))));
            if let Some(waker) = self.waker.take() {
                waker.wake();
            }
```